synapse/storage/schema/delta/55/signing_keys.sql

```diff
@@ -0,0 +1,48 @@
+/* Copyright 2019 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- device signing keys for cross-signing
+CREATE TABLE e2e_device_signing_keys (
+    user_id TEXT NOT NULL,
+    keytype TEXT NOT NULL,
+    keydata TEXT NOT NULL,
+    ts BIGINT NOT NULL
+);
+
+CREATE UNIQUE INDEX e2e_device_signing_keys_idx ON e2e_device_signing_keys(user_id, keytype, ts);
+
+-- devices signatures for cross-signing
+CREATE TABLE e2e_device_signatures (
+    user_id TEXT NOT NULL,
+    key_id TEXT NOT NULL,
+    target_user_id TEXT NOT NULL,
+    target_device_id TEXT NOT NULL,
+    signature TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX e2e_device_signatures_idx ON e2e_device_signatures(user_id, target_user_id, target_device_id);
+
+-- stream of user signature updates
+CREATE TABLE user_signature_stream (
+    stream_id BIGINT NOT NULL,
+    from_user_id TEXT NOT NULL,
+    user_ids TEXT NOT NULL
+);
+
+CREATE INDEX user_signature_stream_idx ON user_signature_stream(stream_id, from_user_id);
```

Comment:
I think I basically cargo-culted the `device_list_stream` from `storage/schema/delta/40/device_list_streams.sql`, which just has an index on `(stream_id, user_id)`.  I don't really understand how stream IDs work, but I can add a unique index on `stream_id`.  `from_user_id` is use in the `WHERE` clause of the query, so it seems to me like it's useful in the index.

Reply:
I wouldn't necessarily take the existing database code as a great example :/.

The query in question is `https://github.com/matrix-org/synapse/commit/8567ff6b4938a477c5b82c5c235124b1ca2e5fbe#diff-21f2f9be6a999959ddb0f9e3769006ccR361` ? Since the `stream_id` part of the query is an open-ended range, it won't be able to make any use of the `user_id` part of the index.